### ISSUE TITLE
Lots Of Compatibility Issues: YUNG'S BETTER CAVES/MINESHAFTS and Other Structures, Biome Makeover Features and Other Stuffs Not Spawning

### ISSUE THREAD

**Issue by Blunder-G** [PolyhedralDev/Terra#167]
There are no new YUNG'S caves. Also, when the Cave Biomes mod is loaded, the generation crashes. I put the crash report: 
[crash-2021-05-08_21.38.51-server.txt](https://github.com/PolyhedralDev/Terra/files/6446471/crash-2021-05-08_21.38.51-server.txt)

Mo 'Structures and Repurposed Structures doesn't occur. Trees and structures of Biome Makeover do not form. YUNG'S BETTER MINESHAFTS mode is also incompatible. Dungeons Of Exile does not work either.

Long story short: Terra removes all worldgen features of other mods. No buildings, trees, mob spawnings...

I have found so many issues for now. Terra is currently inconsistent. However, I still want to use this mod. Do you know if there are any structure datapacks that work with your mod? If so, would you please tell me? 

I hope I was able to help.

**Comment by dfsek** [comment]
> Trees and structures of Biome Makeover do not form.
That would be because the Vanilla biomes that Biome Makeover is "making over" are not generated. The mod simply has nothing to do in a Terra world.

> There are no new YUNG'S caves.
Again, we replace carving. There's nothing for YUNG's caves to generate.

As for general structure compatibility, I am working on an option that will simply inject the parent Vanilla biome's structures into Terra biomes. Mileage may vary depending on how each mod handles structures.

**Comment by deadspidervenom** [comment]
It would be really cool if you carving could be additive instead of replacing, as i like your carving, however Yungs better caves has a lot more too it. ( i am trying to make a private use fantasy theme fabric pack) 

if you are looking to update yours to match their complexity this is the code i do believe they based theirs on (if so that would be cool, i like how more vanilla yours feels)

https://www.curseforge.com/minecraft/mc-mods/worleys-caves

Also can confirm that you get a crash if you try to run terra with "Cave Biomes"
[crash-2021-05-08_21.45.33-server.txt](https://github.com/PolyhedralDev/Terra/files/6446962/crash-2021-05-08_21.45.33-server.txt)

Two mods i think i can confirm compatible are "River Redux" and "Lakeside" Though i am not fulling sure on that

**Comment by dfsek** [comment]
> Two mods i think i can confirm compatible are "River Redux" and "Lakeside" Though i am not fulling sure on that

Neither of those mods will do anything in Terra worlds lol. Sure, they'll *run* but they wont be doing anything. As for Cave Biomes, those features are very simple to implement in config, and I have discussed the crash with SuperCoder and it is something she needs to change on her end.   

As for carving, you can probably just implement most if not all features you want via config. Most of these "tweak" mods should simply be implemented in configs.

**Comment by deadspidervenom** [comment]
I understand most of these changes can be "done in the config" the reason i commented is majority of people dont want to spend the time or dont know how. So you will get a lot of bug reports like this, complaints, and only having a handful of players use your mod due too the forced incompatibility with literally every mod that adds any sort of generation in.

**Comment by Astrashh** [comment]
> Yeah no, when I use a world gen mod I don't want to have to edit a config just to have other mods work with with it. Please don't shoot yourselves in the foot devs.

It's not a matter of having to edit a config to get other mods to work with it, the suggestion is that you can **replicate** features of other mods **through Terra configuration**. If anything, this is a good thing, as when the Terra pack development scene matures, you won't need mods to add to add features to generation, as it might already be integrated into somebody's pack.

> I understand most of these changes can be "done in the config" the reason i commented is majority of people dont want to spend the time or dont know how. So you will get a lot of bug reports like this, complaints, and only having a handful of players use your mod due too the forced incompatibility with literally every mod that adds any sort of generation in.

Just a note on 'forced incompatibility' - other mods are incompatible mainly due to the fact that Terra acts as a *total replacement* of vanilla world generation (which by design allows it to be more maintainable, flexible, and feature-full) compared to other mods & plugins which instead *hook into vanilla generation*. This seems to be a paradigm shift from what people expect, leading to confusion and complaint as to why things don't work the way people want. It's not that it was *intentionally* made to be incompatible, but rather It's due to the very nature of Terra being an independent generator, and the reliance other mods have on vanilla generation.

It's not the sole responsibility of Terra to ensure all other generation mods work with it without any kind of 'compatibility layer' (however the API is being expanded to make that possible), that should be the responsibility of the developers of other mods to allow compatibility with external generators such as Terra, meaning you should complain to the developers of other mods about their mod being incompatible, rather than here.

**Comment by dfsek** [comment]
For that last part, I only partially agree. It isn't solely the other mod developers' responsibility to add compatibility. In some cases we can, and should, add compatibility. Take configured features, for example. We could simply make a first-party addon using our API that injects features from Vanilla biomes into their Terra biome children, or another addon that allows finer-grained control over configured features, injecting them as Terra trees which can be used in config packs.

Simple things like ConfiguredFeatures or structures can easily be made relatively compatible through first-party addons. Things like entire biome emulation are not feasible for us to implement on our end, however, *that's* where mod developers would need to explicitly add compatibility.
